File: streaming/java/src/main/java/compass/main/ExampleClient.java
package compass.main;

import com.google.protobuf.Timestamp;
import compass.harvest.v1alpha1.IngestServiceGrpc;
import compass.harvest.v1alpha1.PositionEventOuterClass.*;
import compass.harvest.v1alpha1.Ingest.*;
import compass.type.geo.Geo;
import io.grpc.*;
import io.grpc.stub.MetadataUtils;

public class ExampleClient {

    public static void main(String[] args) {

        String url = "harvest-grpc.compassiot.cloud";
        String apiKey = "apikey bc8760b6f6f12b8ada9fdfd4503e847f";

        Channel ch = ManagedChannelBuilder.forAddress(url, 443).build();
        IngestServiceGrpc.IngestServiceBlockingStub client = IngestServiceGrpc.newBlockingStub(ch);

        Metadata header = new Metadata();
        Metadata.Key<String> key = Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);
        header.put(key, apiKey);
        client = MetadataUtils.attachHeaders(client, header);

        // See below for more extensive example usage of the partial harvest api
        // if you are NOT using the partial api and just using the start PositionEvent api
        // you may ignore this
        // https://github.com/compass-iot/harvest-api-examples/tree/main/streaming/curl
        PartialEventLatLng partialLatLngEvt = PartialEventLatLng
                .newBuilder()
                .setVehicleId("abc123")
                .setRegion("AUS")
                .setTimestamp(System.currentTimeMillis() / 1000)
                .setLatitude(-33.0)
                .setLongitude(151.0)
                .build();

        PositionEvent evt = PositionEvent
                .newBuilder()
                .setVehicleId("abc123")
                .setVehicleType(VehicleType.CAR)
                .setTransportType(TransportType.PRIVATE)
                .setTimestamp(Timestamp
                        .newBuilder()
                        .setSeconds(System.currentTimeMillis() / 1000)
                        .build())
                .setAcceleration(Acceleration
                        .newBuilder()
                        .setX(1)
                        .setY(2)
                        .setZ(3)
                        .build())
                .setPosition(Position
                        .newBuilder()
                        .setLatlng(Geo.LatLng
                                .newBuilder()
                                .setLatitude(45)
                                .setLongitude(90))
                        .setSpeed(60)
                )
                .setGyro(Gyro
                        .newBuilder()
                        .setRoll(-1)
                        .setPitch(-2)
                        .setYaw(-3))
                .putLabels("make", "toyota")
                .putLabels("something", "nice")
                .build();

        BatchInsertRequest evts = BatchInsertRequest
                .newBuilder()
                .addEvents(evt)
                .build();
        try {
            client.insert(evt);
            client.batchInsert(evts);
            client.insertPartialLatLng(partialLatLngEvt);
        } catch (Exception e) {
            System.err.println(e.getMessage());
            System.exit(1);
        }
        System.out.println("Successfully pushed events via GRPC API!");
    }
}

File: grpc-csharp/ExampleClient.cs
using Compassiot.Compass.V1;
using Compass.Harvest.V1alpha1;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;
using Grpc.Net.Client;

namespace Client
{
    class ExampleClient
    {
        static int Main()
        {
            // Setup client
            string url = "https://harvest.compassiot.cloud";
            GrpcChannel channel = GrpcChannel.ForAddress(url);
            IngestService.IngestServiceClient client = new IngestService.IngestServiceClient(channel);

            // Setup auth
            string apikey = "bc8760b6f6f12b8ada9fdfd4503e847f";
            Metadata headers = new Metadata
            {
                { "Authorization", $"apikey {apikey}" }
            };

            var pe = new PositionEvent
            {
                VehicleId = "abc123",
                VehicleType = VehicleType.Car,
                TransportType = TransportType.Private,
                Timestamp = Timestamp.FromDateTime(DateTime.UtcNow),
                Position = new Position
                {
                    Latlng = new LatLng
                    {
                        Lat = 45,
                        Lng = 90
                    },
                    Speed = 60.0
                },
                Acceleration = new Acceleration
                {
                    X = 1.0,
                    Y = 2.0,
                    Z = 3.0
                },
            };
            pe.Labels.Add(new Dictionary<string, string>
            {
                { "make", "toyota" },
                { "something", "nice" },
            });

            var batch = new BatchInsertRequest
            {
                Events = { pe }
            };

            client.Insert(pe, headers);
            client.BatchInsert(batch, headers);
            Console.WriteLine("Success writing to Harvest");
            return 0;
        }
    }
}
